### PR TITLE
Compressed CSV/JSON support

### LINKED ISSUE #3641: Compressed CSV/JSON Read
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Some data are placed with a compression, e.g., GZip, BZip2. In order to read them with DataFusion, we need to decompress them first.

**Describe the solution you'd like**
Make DataFusion possible to read the compressed CSV/JSON.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Licht-T**
# Which issue does this PR close?
Closes #3641.

 # Rationale for this change
Explained in #3641.

# What changes are included in this PR?
- Add `FileCompressionType` as the text file compression type definition.
- Add GZip/BZip2 converters for `Read/Stream`.
- Add the `COMPRESSION TYPE` SQL token.
- Modify the file type inference method so that it can determine a file type and its compression.

# Are there any user-facing changes?
Yes.

I am not sure which are public APIs, but maybe yes.
> If there are any breaking changes to public APIs, please add the `api change` label.

### CODE REVIEW

**Comment by alamb** [comment]
This looks very cool @Licht-T  -- thank you for the contribution. Hopefully we'll get a chance to review it in the next few days

**Comment by Licht-T** [comment]
Thanks, @alamb.

The test works on my Windows environment but still fails on the Windows CI. I will push a debug code.

**Comment by Licht-T** [comment]
I found `Url::join` cannot handle a drive letter.
```
url: Url { scheme: "file", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/", query: None, fragment: None }, path: "C:/Users/runneradmin/AppData/Local/Temp/.tmpGAMumf/partition-0.json.bz2"
joined url: Url { scheme: "c", cannot_be_a_base: false, username: "", password: None, host: None, port: None, path: "/Users/runneradmin/AppData/Local/Temp/.tmpGAMumf/partition-0.json.bz2", query: None, fragment: None }
```

**Comment by Licht-T** [comment]
Now all green. Ready to get reviewed.

File: datafusion/core/Cargo.toml [hunk 1]
```diff
@@ -57,8 +57,10 @@ unicode_expressions = ["datafusion-physical-expr/regex_expressions", "datafusion
 ahash = { version = "0.8", default-features = false, features = ["runtime-rng"] }
 apache-avro = { version = "0.14", optional = true }
 arrow = { version = "23.0.0", features = ["prettyprint"] }
+async-compression = { version = "0.3.14", features = ["bzip2", "gzip", "futures-io", "tokio"] }
 async-trait = "0.1.41"
 bytes = "1.1"
+bzip2 = "0.4.3"
```
Comment (Dandandan) [inline datafusion/core/Cargo.toml:63]: Any ideas how many dependencies we are adding? I guess some of them are already transitive dependencies through parquet.

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
I took a quick look through this PR and it looks great -- thank you @Licht-T 

Can you please  merge / rebase from master to resolve conflicts so that I can merge it?


Screen Shot 2022-10-03 at 9 50 46 AM


Thanks again@

**Comment by Licht-T** [comment]
@alamb Rebased.

I believe we need some documentation about this, right? I am willing to do that.

**Comment by alamb** [comment]
> I believe we need some documentation about this, right? I am willing to do that.

Yes please @Licht-T  -- that would be super helpful. 

Perhaps https://arrow.apache.org/datafusion/user-guide/sql/ddl.html?highlight=external+table and https://arrow.apache.org/datafusion/user-guide/sql/index.html?highlight=external+table ?

If you are not able to do so, we should at least file a ticket to document the feature

File: datafusion/core/Cargo.toml [hunk 2]
```diff
@@ -101,6 +105,7 @@ ctor = "0.1.22"
 doc-comment = "0.3"
 env_logger = "0.9"
 fuzz-utils = { path = "fuzz-utils" }
+rstest = "0.15.0"
```
Comment (alamb) [inline datafusion/core/Cargo.toml:108]: @Dandandan  what do we think about adding this new testing library?

**Comment by alamb** [comment]
Thanks again @Licht-T

**Comment by Bidek56** [comment]
@Licht-T Were you able to test this change? I am trying these csv options and it simple returns nothing on a valid .gz file.
Thanks

`let csv_options = CsvReadOptions::default().has_header(true).file_compression_type(FileCompressionType::GZIP);
ctx.register_csv("attrib", file, csv_options).await?;
`

**Pull request merged**